src/Widgets/FirmwareUpdateWidget.vala

```diff
@@ -0,0 +1,176 @@
+        var update_to_date_label = new Gtk.Label (_("Up to date")) {
+            valign = Gtk.Align.CENTER
+        };
+        grid.attach (update_to_date_label, 2, 0, 1, 2);
+    }
+
+    private async void update (Device device, Release release) {
+        var path = yield FwupdManager.get_instance ().download_file (release.uri);
+
+        var details = yield FwupdManager.get_instance ().get_details (device, path);
+
+        if (details.caption != null) {
+            show_details_dialog (details);
+        }
+
+        if ((yield FwupdManager.get_instance ().install (device, path)) == true) {
+            if (device.is (DeviceFlag.NEEDS_REBOOT)) {
+                show_reboot_dialog ();
+            } else if (device.is (DeviceFlag.NEEDS_SHUTDOWN)) {
+                show_shutdown_dialog ();
+            }
+        }
+    }
+
+    private void show_details_dialog (Details details) {
+        var message_dialog = new Granite.MessageDialog.with_image_from_icon_name (
+            _("Manual steps required"),
+            details.caption,
+            "application-x-firmware",
+            Gtk.ButtonsType.NONE
+        );
+        message_dialog.transient_for = (Gtk.Window) get_toplevel ();
+
+        var suggested_button = new Gtk.Button.with_label (_("OK"));
+        suggested_button.get_style_context ().add_class (Gtk.STYLE_CLASS_SUGGESTED_ACTION);
+        message_dialog.add_action_widget (suggested_button, Gtk.ResponseType.OK);
+
+        if (details.image != null) {
+            var custom_widget = new Gtk.Image.from_file (details.image);
+            message_dialog.custom_bin.add (custom_widget);
+        }
+
+        message_dialog.badge_icon = new ThemedIcon ("dialog-information");
+        message_dialog.show_all ();
+        message_dialog.run ();
+        message_dialog.destroy ();
+    }
+
+    private void show_reboot_dialog () {
+        var message_dialog = new Granite.MessageDialog.with_image_from_icon_name (
+            _("An update requires a reboot to complete"),
+            _("Reboot now?"),
+            "application-x-firmware",
+            Gtk.ButtonsType.CANCEL
+        );
+        message_dialog.transient_for = (Gtk.Window) get_toplevel ();
+
+        var suggested_button = new Gtk.Button.with_label (_("Reboot"));
+        suggested_button.get_style_context ().add_class (Gtk.STYLE_CLASS_SUGGESTED_ACTION);
+        message_dialog.add_action_widget (suggested_button, Gtk.ResponseType.OK);
+
+        message_dialog.badge_icon = new ThemedIcon ("dialog-information");
+        message_dialog.show_all ();
+        if (message_dialog.run () == Gtk.ResponseType.OK) {
+            LoginManager.get_instance ().reboot ();
+        }
+
+        message_dialog.destroy ();
+    }
+
+    private void show_shutdown_dialog () {
+        var message_dialog = new Granite.MessageDialog.with_image_from_icon_name (
+            _("An update requires the system to shutdown to complete"),
+            _("Shutdown now?"),
```

Comment:
Probably should use the same copy from the shutdown dialog in indicator session:
```suggestion
            _("This will close all open applications and turn off this device."),
```